src/objects/set.rs

```diff
@@ -3,152 +3,155 @@
 
 use std::{hash, collections};
 use ffi;
-use pyptr::Py;
-use python::Python;
-use conversion::ToPyObject;
-use objects::PyObject; //, PyIterator};
+use pyptr::PyPtr;
+use python::{Python, ToPythonPointer};
+use conversion::{ToPyObject, IntoPyObject};
+use objects::{PyObject, PyIterator};
 use err::{self, PyResult, PyErr};
+use pptr;
+use token::{PyObjectMarker, PythonObjectWithGilToken};
+use objectprotocol::ObjectProtocol;
 
 
 /// Represents a Python `set`
-pub struct PySet;
+pub struct PySet<'p>(pptr<'p>);
 /// Represents a  Python `frozenset`
-pub struct PyFrozenSet;
+pub struct PyFrozenSet<'p>(pptr<'p>);
 
-pyobject_newtype!(PySet, PySet_Check, PySet_Type);
-pyobject_newtype!(PyFrozenSet, PyFrozenSet_Check, PyFrozenSet_Type);
+pyobject_nativetype!(PySet, PySet_Check, PySet_Type);
+pyobject_nativetype!(PyFrozenSet, PyFrozenSet_Check, PyFrozenSet_Type);
 
-impl PySet {
+impl<'p> PySet<'p> {
     /// Creates a new set.
     ///
     /// May panic when running out of memory.
-    pub fn new<T: ToPyObject>(py: Python, elements: &[T]) -> PySet {
-        let list = elements.to_py_object(py);
+    pub fn new<T: ToPyObject>(py: Python<'p>, elements: &[T]) -> PySet<'p> {
+        let list = elements.to_object(py);
         unsafe {
-            err::cast_from_owned_ptr_or_panic(py, ffi::PySet_New(list.as_ptr()))
+            let ptr = ffi::PySet_New(list.as_ptr());
```

Comment:
do you know if PySet_New increases ref counter? if it just takes ownership then this is bug.